src/tracer/propagation/cloud_trace_propagator.cpp

```diff
@@ -0,0 +1,187 @@
+              baggage.insert(BaggageProtobufMap::value_type(
+                  ToLower(
+                      opentracing::string_view{key.data() + PrefixBaggage.size(),
+                                               key.size() - PrefixBaggage.size()}),
+                  value));
+            }
+            return {};
+          });
+      if (!result) {
+        return opentracing::make_unexpected(result.error());
+      }
+      return parent_header_found;
+}
+
+opentracing::expected<void> CloudTracePropagator::ParseCloudTrace(
+    opentracing::string_view s, TraceContext& trace_context) const noexcept {
+  if (s.size() < Num128BitHexDigits) {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+  size_t offset = 0;
+
+  // default sampled to on (this comes from the ;o=1 part of
+  // x-cloud-trace-context; if it is not set we will default to sampling
+  // this request)
+  trace_context.trace_flags = SetTraceFlag<SampledFlagMask>(trace_context.trace_flags, true);
+
+  // trace-id
+  auto error_maybe = NormalizedHexToUint128(
+      opentracing::string_view{s.data() + offset, Num128BitHexDigits},
+      trace_context.trace_id_high, trace_context.trace_id_low);
+  if (!error_maybe) {
+    return error_maybe;
+  }
+
+  offset += Num128BitHexDigits;
+  if (s.size() < Num128BitHexDigits + 2) {
+    // only a short form trace ID has been given (not a "trace id/span id")
+    return {};
+  }
+
+  if (s[offset] != '/') {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+  ++offset;
+
+  char parent_id [21];
+  int span_id_length;
+  for(span_id_length = 0; span_id_length<20 && offset<s.length(); span_id_length++) {
+    if(s[offset]==';' || s[offset]=='\0') {
+      break;
+    }
+
+    parent_id[span_id_length] = s[offset];
+    ++offset;
+  }
+  parent_id[span_id_length] = '\0';
+
+  // parent-id
+  char * pEnd = parent_id;
+  trace_context.parent_id = std::strtoull(pEnd, nullptr, 10);
+
+  if (s.size() - offset < 4) {
+    // only a "trace ID/span ID" has been given (not a "trace id/span id;o=[0-1]")
+    return {};
+  }
+
+  if(opentracing::string_view(s.begin() + offset, 3) != opentracing::string_view(";o=")) {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+
+  offset += 3;
+
+  // trace-flags
+  if (s[offset] == '0') {
+    // don't sample
+    trace_context.trace_flags = SetTraceFlag<SampledFlagMask>(trace_context.trace_flags, false);
+  }
+
+  return {};
+}
+
+size_t CloudTracePropagator::SerializeCloudTrace(const TraceContext& trace_context,
+                           char* s) const noexcept {
+  // trace-id
+  char trace_id[Num64BitHexDigits + Num64BitHexDigits + 1];
+  Uint64ToHex(trace_context.trace_id_high, trace_id);
+  Uint64ToHex(trace_context.trace_id_low, trace_id + Num64BitHexDigits);
+  *(trace_id + Num64BitHexDigits + Num64BitHexDigits) = '\0';
```

Comment:
You don't need to copy these twice -- just serialized the trace ids directly to s, then use snprintf for the remaining part. Or avoid the Uint64ToHex calls and use only the printf formatting

Reply:
thanks, done